src/librustc_mir/transform/uniform_array_move_out.rs

```diff
@@ -151,3 +152,177 @@ impl<'a, 'tcx> UniformArrayMoveOutVisitor<'a, 'tcx> {
+        let mut patch = MirPatch::new(mir);
+        {
+            let mut visitor = RestoreDataCollector {
+                locals_use: IndexVec::new(),
+                candidates: vec![],
+            };
+            visitor.locals_use.resize(mir.local_decls.len(), LocalUse::new());
+            visitor.visit_mir(mir);
+
+            for candidate in &visitor.candidates {
+                let statement = &mir[candidate.block].statements[candidate.statement_index];
+                if let StatementKind::Assign(ref dst_place, ref rval) = statement.kind {
+                    if let Rvalue::Aggregate(box AggregateKind::Array(_), ref items) = *rval {
+                        let items : Vec<_> = items.iter().map(|item| {
+                            if let Operand::Move(Place::Local(local)) = item {
+                                let local_use = &visitor.locals_use[*local];
+                                let opt_index_and_place = try_get_item_source(local_use, mir);
+                                if local_use.use_count == 2 && opt_index_and_place.is_some() {
+                                    let (index, src_place) = opt_index_and_place.unwrap();
+                                    return Some((local_use, index, src_place));
+                                }
+                            }
+                            None
+                        }).collect();
+
+                        let opt_src_place = items.first().and_then(|x| *x).map(|x| x.2);
+                        let opt_size = opt_src_place.and_then(|src_place| {
+                            let src_ty = src_place.ty(mir, tcx).to_ty(tcx);
+                            if let ty::TyArray(_, ref size_o) = src_ty.sty {
+                                size_o.val.to_const_int().and_then(|v| v.to_u64())
+                            } else {
+                                None
+                            }
+                        });
+
+                        check_and_patch(*candidate, &items, opt_size, &mut patch, dst_place);
+                    }
+                }
+            }
+        }
+        patch.apply(mir);
+    }
+}
+
+fn try_get_item_source<'a, 'tcx>(local_use: &LocalUse,
+                                 mir: &'a Mir<'tcx>) -> Option<(u32, &'a Place<'tcx>)> {
+    if let Some(location) = local_use.first_use {
+        let block = &mir[location.block];
+        if block.statements.len() > location.statement_index {
+            let statement = &block.statements[location.statement_index];
+            if let StatementKind::Assign(
+                Place::Local(_),
+                Rvalue::Use(Operand::Move(Place::Projection(box PlaceProjection{
+                    ref base, elem: ProjectionElem::ConstantIndex{
+                        offset, min_length: _, from_end: false}})))) = statement.kind {
+                return Some((offset, base))
+            }
+        }
+    }
+    None
+}
+
+fn check_and_patch<'tcx>(candidate: Location,
```

Comment:
Nit: I'd like to see some more comments. =) 

For example, `items: &Vec<Option<(&LocalUse, u32, &Place<'tcx>)>>` is a pretty dense type. It'd be cool to explain in terms of your example what values we should expect for all those fields etc.